src/librustc_trans/trans/base.rs

```diff
@@ -2528,25 +2528,44 @@ pub fn write_metadata(cx: &SharedCrateContext, krate: &ast::Crate) -> Vec<u8> {
 
     let encode_parms = crate_ctxt_to_encode_parms(cx, encode_inlined_item);
     let metadata = encoder::encode_metadata(encode_parms, krate);
-    let mut compressed = encoder::metadata_encoding_version.to_vec();
-    compressed.push_all(&flate::deflate_bytes(&metadata));
-    let llmeta = C_bytes_in_context(cx.metadata_llcx(), &compressed[..]);
-    let llconst = C_struct_in_context(cx.metadata_llcx(), &[llmeta], false);
-    let name = format!("rust_metadata_{}_{}",
-                       cx.link_meta().crate_name,
-                       cx.link_meta().crate_hash);
-    let buf = CString::new(name).unwrap();
-    let llglobal = unsafe {
-        llvm::LLVMAddGlobal(cx.metadata_llmod(), val_ty(llconst).to_ref(),
-                            buf.as_ptr())
-    };
-    unsafe {
-        llvm::LLVMSetInitializer(llglobal, llconst);
-        let name = loader::meta_section_name(&cx.sess().target.target);
-        let name = CString::new(name).unwrap();
-        llvm::LLVMSetSection(llglobal, name.as_ptr())
+
+    let need_metadata_module = cx.sess().crate_types.borrow().iter().any(|ty| {
+        *ty == config::CrateTypeDylib
+    });
+
+    if !need_metadata_module {
+        return (metadata, None);
+    } else {
+        let (metadata_llcx, metadata_llmod) = unsafe {
+            create_context_and_module(cx.sess(), "metadata")
+        };
+        
+        let mut compressed = encoder::metadata_encoding_version.to_vec();
+        compressed.push_all(&flate::deflate_bytes(&metadata));
+        let llmeta = C_bytes_in_context(metadata_llcx, &compressed[..]);
+        let llconst = C_struct_in_context(metadata_llcx, &[llmeta], false);
+        let name = format!("rust_metadata_{}_{}",
+                           cx.link_meta().crate_name,
+                           cx.link_meta().crate_hash);
+        let buf = CString::new(name).unwrap();
+        let llglobal = unsafe {
+            llvm::LLVMAddGlobal(metadata_llmod, val_ty(llconst).to_ref(),
+                                buf.as_ptr())
+        };
+        unsafe {
+            llvm::LLVMSetInitializer(llglobal, llconst);
+            let name = loader::meta_section_name(&cx.sess().target.target);
+            let name = CString::new(name).unwrap();
+            llvm::LLVMSetSection(llglobal, name.as_ptr())
+        }
+
+        let metadata_module = Some(ModuleTranslation {
+            llcx: metadata_llcx,
+            llmod: metadata_llmod,
+        });
+
+        return (metadata, metadata_module);
```

Comment:
Could this all be tabbed left by one after the `return (metadata, None)` case above? Helps rightward drift a bit.